### PR TITLE
Refactor aggregator and evaluationManager

### PR DESCRIPTION
**Pull request by qibobo**
Just refactor the two class: aggregator and evaluationManager and their unit tests. The code in /cmd folder is still the old one. 
Create this pull request for your review to ensure that I am in the right way.

### CODE REVIEW

File: src/autoscaler/eventgenerator/aggregator/aggregator.go [hunk 1]
```diff
@@ -1,119 +1,47 @@
...
+type GetAppMonitors func() []*model.AppMonitor
 type Aggregator struct {
 	logger                    lager.Logger
 	doneChan                  chan bool
 	appChan                   chan *model.AppMonitor
-	policyPoller              *PolicyPoller
-	metricPollerArray         []*MetricPoller
-	appMetricDatabase         db.AppMetricDB
-	evaluationManager         *generator.AppEvaluationManager
+	getAppMonitors            GetAppMonitors
```
Comment (fraenkel) [inline src/autoscaler/eventgenerator/aggregator/aggregator.go]: This should be GetPolicies()

File: src/autoscaler/eventgenerator/aggregator/aggregator.go [hunk 2]
```diff
@@ -1,119 +1,47 @@
...
-		doneChan:          make(chan bool),
-		appChan:           appMonitorChan,
-		metricPollerArray: []*MetricPoller{},
-		appMetricDatabase: appMetricDatabase,
-		evaluationManager: evaluationManager,
-		cclock:            clock,
+		logger:         logger.Session("Aggregator"),
+		doneChan:       make(chan bool),
+		appChan:        appMonitorChan,
+		getAppMonitors: getAppMonitors,
```
Comment (fraenkel) [inline src/autoscaler/eventgenerator/aggregator/aggregator.go]: GetPolicies

File: src/autoscaler/eventgenerator/aggregator/aggregator.go [hunk 3]
```diff
@@ -1,119 +1,47 @@
...
+		cclock:         clock,
 		aggregatorExecuteInterval: aggregatorExecuteInterval,
 		appMonitorArray:           []*model.AppMonitor{},
 	}
-	aggregator.policyPoller = NewPolicyPoller(logger, clock, policyPollerInterval, policyDatabase, aggregator.ConsumePolicy, aggregator.appChan)
-	client := cfhttp.NewClient()
-	client.Transport.(*http.Transport).MaxIdleConnsPerHost = metricPollerCount
 
-	var i int
-	for i = 0; i < metricPollerCount; i++ {
```
Comment (fraenkel) [inline src/autoscaler/eventgenerator/aggregator/aggregator.go]: The MetricPoller should stay for now. I will clean that up in a future PR

File: src/autoscaler/eventgenerator/aggregator/aggregator.go [hunk 4]
```diff
@@ -1,119 +1,47 @@
...
-		poller := NewMetricPoller(metricCollectorUrl, logger, aggregator.appChan, aggregator.ConsumeAppMetric, client)
-		aggregator.metricPollerArray = append(aggregator.metricPollerArray, poller)
-	}
 	return aggregator
 }
 
-func (a *Aggregator) ConsumePolicy(policyMap map[string]*model.Policy, appChan chan *model.AppMonitor) {
```
Comment (fraenkel) [inline src/autoscaler/eventgenerator/aggregator/aggregator.go]: Parts of this need to stay so you can convert Polciies -> []*model.AppMonitor

File: src/autoscaler/eventgenerator/aggregator/aggregator.go [hunk 5]
```diff
@@ -1,119 +1,47 @@
...
-			})
-			triggerArrayMap[triggerKey] = triggerArray
-		}
-	}
-
-	a.setAppMonitors(appMonitorArrayTmp)
-	a.evaluationManager.SetTriggers(triggerArrayMap)
-}
-
-func (a *Aggregator) ConsumeAppMetric(appMetric *model.AppMetric) {
```
Comment (fraenkel) [inline src/autoscaler/eventgenerator/aggregator/aggregator.go]: Keep this for now.

File: src/autoscaler/eventgenerator/aggregator/aggregator.go [hunk 6]
```diff
@@ -1,119 +1,47 @@
...
-	}
-	err := a.appMetricDatabase.SaveAppMetric(appMetric)
-	if err != nil {
-		a.logger.Error("save appmetric to database failed", err, lager.Data{"appmetric": appMetric})
-	}
-}
-
 func (a *Aggregator) Start() {
-	a.policyPoller.Start()
-	for _, metricPoller := range a.metricPollerArray {
```
Comment (fraenkel) [inline src/autoscaler/eventgenerator/aggregator/aggregator.go]: Keep

File: src/autoscaler/eventgenerator/aggregator/aggregator.go [hunk 7]
```diff
@@ -1,119 +1,47 @@
...
-	}
-	go a.startWork()
-
+	go a.start()
 	a.logger.Info("started")
 }
 
 func (a *Aggregator) Stop() {
-	a.policyPoller.Stop()
-	for _, metricPoller := range a.metricPollerArray {
```
Comment (fraenkel) [inline src/autoscaler/eventgenerator/aggregator/aggregator.go]: Keep

File: src/autoscaler/eventgenerator/aggregator/aggregator.go [hunk 8]
```diff
@@ -124,25 +52,18 @@ func (a *Aggregator) setAppMonitors(appMonitors []*model.AppMonitor) {
...
+func (a *Aggregator) start() {
 	ticker := a.cclock.NewTicker(a.aggregatorExecuteInterval)
 	defer ticker.Stop()
 	for {
 		select {
 		case <-a.doneChan:
 			return
 		case <-ticker.C():
-			a.addToAggregateChannel()
+			a.setAppMonitors(a.getAppMonitors())
```
Comment (fraenkel) [inline src/autoscaler/eventgenerator/aggregator/aggregator.go]: monitors := getAppMonitors(a.GetPolicies())

File: src/autoscaler/eventgenerator/generator/appEvaluationManager.go [hunk 9]
```diff
@@ -1,17 +1,14 @@
...
-	"code.cloudfoundry.org/cfhttp"
 	"code.cloudfoundry.org/clock"
 	"code.cloudfoundry.org/lager"
-	"net/http"
 	"sync"
 	"time"
 )
 
-type ConsumeAppMonitorMap func(map[string][]*model.Trigger, chan []*model.Trigger)
+type GetTriggers func() map[string][]*model.Trigger
```
Comment (fraenkel) [inline src/autoscaler/eventgenerator/generator/appEvaluationManager.go]: GetPolicies()

File: src/autoscaler/eventgenerator/generator/appEvaluationManager.go [hunk 10]
```diff
@@ -20,32 +17,24 @@ type AppEvaluationManager struct {
 	doneChan         chan bool
 	triggerChan      chan []*model.Trigger
 	triggers         map[string][]*model.Trigger
-	evaluatorArray   []*Evaluator
+	getTriggers      GetTriggers
```
Comment (fraenkel) [inline src/autoscaler/eventgenerator/generator/appEvaluationManager.go]: GetPolicies()

File: src/autoscaler/eventgenerator/generator/appEvaluationManager.go [hunk 11]
```diff
@@ -20,32 +17,24 @@ type AppEvaluationManager struct {
...
 		logger:           logger.Session("AppEvaluationManager"),
 		cclock:           cclock,
 		doneChan:         make(chan bool),
 		triggerChan:      triggerChan,
 		triggers:         map[string][]*model.Trigger{},
-		evaluatorArray:   []*Evaluator{},
-	}
-	client := cfhttp.NewClient()
-	client.Transport.(*http.Transport).MaxIdleConnsPerHost = evaluatorCount
-	for i := 0; i < evaluatorCount; i++ {
```
Comment (fraenkel) [inline src/autoscaler/eventgenerator/generator/appEvaluationManager.go]: Keep all this

File: src/autoscaler/eventgenerator/generator/appEvaluationManager.go [hunk 12]
```diff
@@ -20,32 +17,24 @@ type AppEvaluationManager struct {
...
-		evaluator := NewEvaluator(logger, client, scalingEngineUrl, triggerChan, database)
-		manager.evaluatorArray = append(manager.evaluatorArray, evaluator)
+		getTriggers:      getTriggers,
 	}
 	return manager
 }
 func (a *AppEvaluationManager) Start() {
-	for _, evaluator := range a.evaluatorArray {
-		evaluator.Start()
```
Comment (fraenkel) [inline src/autoscaler/eventgenerator/generator/appEvaluationManager.go]: Keep

File: src/autoscaler/eventgenerator/generator/appEvaluationManager.go [hunk 13]
```diff
@@ -54,31 +43,25 @@ func (a *AppEvaluationManager) Stop() {
...
+func (a *AppEvaluationManager) start() {
 	ticker := a.cclock.NewTicker(a.evaluateInterval)
 	defer ticker.Stop()
 	for {
 		select {
 		case <-a.doneChan:
 			return
 		case <-ticker.C():
-			a.evaluate()
+			a.setTriggers(a.getTriggers())
```
Comment (fraenkel) [inline src/autoscaler/eventgenerator/generator/appEvaluationManager.go]: triggers := a.getTriggers(a.GetPolicies())

getTriggers is lifted from the Aggregator.

**Comment by qibobo** [comment]
reset code and update.

**Comment by fraenkel** [comment]
Much better... I will fix up the merge conflict and remove and reduce the locking that is there.

**Pull request merged**